app/models/manageiq/providers/cloud_manager/auth_key_pair.rb

```diff
@@ -33,18 +33,13 @@ def self.create_key_pair_queue(userid, ext_management_system, options = {})
 
   def self.create_key_pair(ems_id, options)
     raise ArgumentError, _("ems cannot be nil") if ems_id.nil?
+
     ext_management_system = ExtManagementSystem.find(ems_id)
-    raise ArgumentError, _("ems cannot be found") if ext_management_system.nil?
 
     klass = ext_management_system.class_by_ems(:AuthKeyPair)
-    # TODO(maufart): add cloud_tenant to database table?
-    created_key_pair = klass.raw_create_key_pair(ext_management_system, options)
-    klass.create(
-      :name        => created_key_pair.name,
-      :fingerprint => created_key_pair.fingerprint,
-      :resource    => ext_management_system,
-      :auth_key    => created_key_pair.private_key
-    )
+
+    key_pair_opts = klass.raw_create_key_pair(ext_management_system, options)
```

Comment:
so `raw_create_key_pair` does not actually create a key pair but rather returns options that then are going to be passed into `create` to actually create the key pair?

(sorry for the snark) - Is this a change in the underlying `raw_create_key_pair` code, or was it just plain broken before?

Reply:
It creates the keypair on the native provider, not in the database.  This is the same behavior as all raw_create_* methods, this is different in that it also creates the database record because if the cloud provider generates the private key it is only possible to be gotten in the API response not in the subsequent refresh so we have to create the record here.

If raw_create_key_pair created the database record also this would be duplicated by every provider, so raw_* is only the provider specific bits.